### ISSUE TITLE
JSON params not changing on multiple posts to controller action

### ISSUE THREAD

**Issue by dereksweet** [rails/rails#31643]
Seem to have an issue where a single test that makes multiple posts to the same controller action, with different params, does not seem to receive the subsequent changes in params. For example, if I have a very simple rails project with just a single controller, with a single action, and a single test that tests that action, the params do not seem to change between posts to that action. 

Here is the controller:
```
class TestController < ApplicationController
  def index
    render json: { my_string: params[:string_param], my_int: params[:int_param] }
  end
end
```

Here is the spec:
```
require 'rails_helper'

RSpec.describe TestController, type: :controller do

  it 'should return the params that are passed in' do
    request.content_type = 'application/json'

    post :index, params: { string_param: 'test1', int_param: 1 }

    expect(response.status).to eq(200)
    json_response = JSON.parse(response.body)
    expect(json_response['my_string']).to eq('test1')
    expect(json_response['my_int']).to eq(1)

    post :index, params: { string_param: 'test2', int_param: 2 }

    expect(response.status).to eq(200)
    json_response = JSON.parse(response.body)
    expect(json_response['my_string']).to eq('test2')
    expect(json_response['my_int']).to eq(2)
  end

end
```

This test will fail on `expect(json_response['my_string']).to eq('test2')` because the controller will return the params passed in on the first post and `my_string` will be `"test1"`, not `"test2"`.

Alternatively your spec could look like this and you get the same behavior:
```
require 'rails_helper'

RSpec.describe TestController, type: :controller do

  it 'should return the params that are passed in' do
    post :index, params: { string_param: 'test1', int_param: 1 }, as: :json

    expect(response.status).to eq(200)
    json_response = JSON.parse(response.body)
    expect(json_response['my_string']).to eq('test1')
    expect(json_response['my_int']).to eq(1)

    post :index, params: { string_param: 'test2', int_param: 2 }, as: :json

    expect(response.status).to eq(200)
    json_response = JSON.parse(response.body)
    expect(json_response['my_string']).to eq('test2')
    expect(json_response['my_int']).to eq(2)
  end

end
```

I have tried using `format: :json` instead of `as: :json`, but then my integers are converted into strings and the spec will fail on the line `expect(json_response['my_int']).to eq(1)` because `my_int` is coming back as `"1"`, not `1`.

I have verified with `Rails.logger.info(params.inspect)` in the controller action that, indeed, the params on the 2nd post are the same as the ones passed in on the first post. 

Anyone know what's going on here? Is this a bug with Rails or RSpec? Should I be making an issue over a the RSpec gem repo instead of Rails?

**Comment by mrstebo** [comment]
I assume the test works if you split the tests up? would be interesting to see what is going on... 
My head is telling me that maybe it could be caching, but probably not. I'll have to look a bit deeper into how controller tests are set up in Rspec

**Comment by dereksweet** [comment]
Yes that is correct. The tests would pass if they weren't within the same `it` block.

**Comment by mrstebo** [comment]
Okay, just ran a test. If I used `post` then it would mimic the behaviour that you have outlined. However, the route that is generated when you run `rails g controller Test index` produces a route in **routes.rb** - `get 'test/index'`

So if I run your tests changing `post` to `get`, then the tests work fine.

Could you verify what is in your **routes.rb** file?

**Comment by dereksweet** [comment]
Yes I noticed that to, the problem only seems to affect `post`, `put`, and `patch`.

**Comment by mrstebo** [comment]
This is probably more to do with rspec. See [here](https://stackoverflow.com/questions/7850650/making-two-requests-to-the-same-controller-in-rails-integrations-specs) for a Stack Overflow post regarding a similar issue.

**Comment by dereksweet** [comment]
Thanks for your comments. I'll investigate if the problems happens with TestUnit and report it over at Rspec if appropriate.

**Comment by amccann** [comment]
Commenting because I came across this issue when googling about similar behaviour in 5.0.7 and want to point the next person in the right direction.

--

The issue exists in `5-0-stable` and `5-1-stable`, but is now fixed in `5-2-stable` and the release `5.2.1`.

The following PR fixes the issue
https://github.com/rails/rails/pull/32673

**Comment by marekciupak** [comment]
> The following PR fixes the issue
> #32673

@andrewjdmccann Thank you for writing this comment. You just saved me a lot of time.

A temporary fix for people who can not upgrade Rails to 5.2:

```diff
  post :index, params: { string_param: 'test1', int_param: 1 }, as: :json
+ reset_raw_post_data
  post :index, params: { string_param: 'test2', int_param: 2 }, as: :json
+ reset_raw_post_data
  post :index, params: { string_param: 'test3', int_param: 3 }, as: :json
```

```ruby
# TODO: Remove it after upgrading Rails to version 5.2.
# Resets RAW_POST_DATA that is derived from the rack.input header, which changes with each test request.
# It needs to be cleared, or all requests in the same test will see the value from the first request.
# This has been fixed in Rails 5.2: https://github.com/rails/rails/pull/32673.
def reset_raw_post_data
  @request.env.delete("RAW_POST_DATA")
end
```
